### DISCUSSION TITLE
Better TorGuard Integration

### DISCUSSION THREAD

**Discussion by torguardops** [haugene/docker-transmission-openvpn#2403] [General]
Hello @haugene - we love your project and would love to find a way to better integrate TorGuard so that users do not have to depend on outdated configs. I could not find another good way to contact you directly so I am including my email here for you: ops@torguard.net

**Comment by pkishino** [comment]
@torguardops hello,
pretty cool that a provider is reaching out to us.
I:'ll let @haugene respond, but I would suggest a simple config script that instead of fetching from our main provider repo, would fetch directly from you instead ?

**Comment by haugene** [comment]
Hey @torguardops and thank you for your interest in the project. I'll send you an email as well, but the standard way to do this is to set up the container to dynamically fetch configs directly from you instead of through the [provider repo](https://github.com/haugene/vpn-configs-contrib) (or some fork of it) like TorGuard is today.

The container supports three main types of configuration, and that can be classified as `internal, external and custom`. The `internal` ones are maintained in the main image and is in the form of a script that downloads a bundle of configs and determines which to use. The `custom` config setup is used to mount a config file directly off the host machine and lastly the `external` providers are the ones that are checked into the config repository and needs to be manually updated there by someone in the community, which usually leads to a certain percentage of them being outdated at any point. A bit more on that here: https://haugene.github.io/docker-transmission-openvpn/supported-providers/

What needs to be done is to create a script called `configure-openvpn.sh` and put it in a `torguard` folder within this repo and it should work automatically. The default behaviour is to use the internal provider if it exists and then fetch the external one if it isn't. Then next level of integration is to also provide an `update-port.sh` script that will be called after the VPN connection is established. That can do what needs to be done to get an open port and then we can set that as the port to use by Transmission.

Let me know what you think.

**Reply by torguardops** [reply]
I replied to your email and look forward to helping set something up @haugene! I am in discussions with my team now to see what kind of solution we can work out and will keep in contact with you by email.

You explained it quite well here and I replied to your email before I had a chance to read your reply but we are working out the most secure way that this can be done from a user standpoint API wise for the port forwarding.

Something can be worked out for sure!

Thanks again!

**Comment by haugene** [comment]
Hello again @torguardops. Just wanted to let you know that I have updated the configs based on the .zip file of configs you sent me. See PR https://github.com/haugene/vpn-configs-contrib/pull/160

There were some errors though. The configs don't run out of the box on linux as there's an unrecognized option there. Others failed on dns lookup or timeout. I ran a test of all the configs and attached the results in the PR. My test script doesn't guarantee that the config would work because it's a generic script that assumes that it doesn't have an account and is satisfied with getting to an `AUTH_FAILED` status. Not all configs got that far.

I ended up removing 13 configs to avoid a bad user experience of having to bang their heads on these. I listed them in the PR, interested to hear if you see some errors on the on your side or if they're working as expected otherwise.

**Reply by torguardops** [reply]
Thanks @haugene ! I passed it along so they can check into it.

Our end goal is to set up a system where users are able to retrieve a current config from the server based on the location name and this way they will always be up to date. This will output a config that the script can "fill in the blank" to place the login variables.

The developers are also working on a system that works with port forwarding configurations being pulled through the same method using a sort of "key" they are given from the TorGuard panel once they configure the port forward.

To give an example: User goes to the panel and sets up a port forward for a specific location and port. They are given a unique identifier that can be used as a variable inside the container. The script inside the container will be able to pull the config for that key but for security reasons it will not contain authentication data.

Once the config is retrieved and the auth variables filled in you are then able to connect. This way it works for those who wish to port forward and those who just want to use a config by location.

Hope that makes sense and I will keep this thread updated as our developers give me more information.

**Comment by AhnFire** [comment]
Several of the TorGuard US entries are not working. I have tried to make a pull request to try and add a new server. I'm not sure if it will work, as I do not know how to test it (maybe I can fork the entire repository into my own user and try to deploy to docker from there?). Sorry I am a docker newbie. Been using Portainer.

I also got errors with a script and started a discussion under the config repo, after specifying my fork as an environment variable.

**Reply by HelloMihai** [reply]
Hi, any update on this?  Im a long time `TorguardVPN` user and would love a seamless integration in this docker.

**Reply by torguardops** [reply]
It has been a long while since I followed up on all of this and I am going to be working on it shortly. Thanks @AhnFire  for your work on this too.

I have also switched my main account so any posts from [FixedBit](https://github.com/fixedbit) are me.
